src/Provider/AbstractProvider.php

```diff
@@ -291,51 +288,63 @@ public function getAccessToken($grant = 'authorization_code', array $params = []
 
         $requestParams = $grant->prepRequestParams($defaultParams, $params);
 
-        try {
-            $client = $this->getHttpClient();
-            switch (strtoupper(static::ACCESS_TOKEN_METHOD)) {
-                case 'GET':
-                    // @codeCoverageIgnoreStart
-                    // No providers included with this library use get but 3rd parties may
-                    $httpResponse = $client->get(
-                        $this->urlAccessToken(),
-                        $this->getHeaders(),
-                        $requestParams
-                    );
-                    $response = (string) $httpResponse->getBody();
-                    break;
-                    // @codeCoverageIgnoreEnd
-                case 'POST':
-                    $httpResponse = $client->post(
-                        $this->urlAccessToken(),
-                        $this->getHeaders(),
-                        $requestParams
-                    );
-                    $response = (string) $httpResponse->getBody();
-                    break;
+        switch (strtoupper(static::ACCESS_TOKEN_METHOD)) {
+            case 'GET':
                 // @codeCoverageIgnoreStart
-                default:
-                    throw new \InvalidArgumentException('Neither GET nor POST is specified for request');
+                // No providers included with this library use get but 3rd parties may
+                $options['query'] = $requestParams;
+                break;
                 // @codeCoverageIgnoreEnd
-            }
-        } catch (HttpAdapterException $e) {
-            if (!$e->hasResponse()) {
-                throw $e;
-            }
-            $response = (string) $e->getResponse()->getBody();
+            case 'POST':
+                $options['form_params'] = $requestParams;
+                break;
+            // @codeCoverageIgnoreStart
+            default:
+                throw new \InvalidArgumentException('Neither GET nor POST is specified for request');
+            // @codeCoverageIgnoreEnd
         }
 
-        $response = $this->parseResponse($response);
+        $request  = $this->getGuzzleRequest(static::ACCESS_TOKEN_METHOD, $this->urlAccessToken());
+        $response = $this->getResponse($request);
         $response = $this->prepareAccessTokenResult($response);
 
         return $grant->handleResponse($response);
     }
 
     /**
+     * Get a request instance, configured with Guzzle.
+     *
+     * Available options are described here: http://docs.guzzlephp.org/en/latest/request-options.html
+     *
+     * @param  string $method
+     * @param  string $url
+     * @parma  array  $options
+     * @return GuzzleHttp\Psr7\Request
+     */
+    protected function getGuzzleRequest($method, $url, array $options = [])
+    {
+        return $this->getHttpClient()->requestFactory($method, $url, $options);
```

Comment:
@mtdowling here we use theoretical Guzzle request factory. Note that the interface is not public, the fact that this is Guzzle doesn't matter.